lsp-mode.el

```diff
@@ -3913,24 +3913,27 @@ Applies on type formatting."
 
 ;; links
 (defun lsp--document-links ()
-  (lsp-request-async
-   "textDocument/documentLink"
-   `(:textDocument ,(lsp--text-document-identifier))
-   (lambda (links)
-     (lsp--remove-overlays 'lsp-link)
-     (seq-do
-      (-lambda ((link &as &hash "range"))
-        (-doto (make-button (lsp--position-to-point (gethash "start" range))
-                            (lsp--position-to-point (gethash "end" range))
-                            'action (lsp--document-link-keymap link)
-                            'keymap (let ((map (make-sparse-keymap)))
-                                      (define-key map [M-return] 'push-button)
-                                      (define-key map [mouse-2] 'push-button)
-                                      map)
-                            'help-echo "mouse-2, M-RET: Visit this link")
-          (overlay-put 'lsp-link t)))
-      links))
-   :mode 'tick))
```

Comment:
I think that all this is a regression from the lsp-request-async refactoring. In the past, we were wrapping the callback in lsp--create-async-callback for `tick/alive` and we were doing with-current-buffer in it. Would you mind to fix the issue it that way to avoid the code duplication?

Reply:
sure, should have checked the commit history first. Thanks for the info

Reply:
I've now had a look at this and I agree that `'tick` should use `(with-current-buffer ...)`, though I guess that most current users of `'tick` should probably switch to `'unchanged` anyway? Legitimate uses of a buffer-changing `'tick` implementation would probably be `lsp--on-type-formatting` and `lsp--before-save`, but `lsp--document-links`, `lsp-hover` and `lsp--document-highlight` not so much. Do you agree?

Reply:
Yes.

Reply:
Since all modes cancel the request if the current buffer is killed, I'd just do the `(with-current-buffer ...)` unconditionally, or would you prefer not to needlessly wrap the callback if mode is `'current` or `'unchanged`?

Reply:
> Since all modes cancel the request if the current buffer is killed, 

`detached` is supposed not to do that?

Reply:
oh yes, I was looking at the values listed in the `(pcase mode`, not at the docstring. So, I'll just wrap it for `alive/tick` then as you originally proposed

Reply:
do you see any issues with the change? I've been using it for about a week, seemed to work for me so far